### PR TITLE
[IR] Fix GEP offset computations for vector GEPs

### PR DESCRIPTION
**Pull request by jasilvanus**
Vectors are always bit-packed and don't respect the elements' alignment
requirements. This is different from arrays. This means offsets of vector GEPs
need to be computed differently than offsets of array GEPs.

This PR fixes many places that rely on an incorrect pattern
that always relies on `DL.getTypeAllocSize(GTI.getIndexedType())`.
We replace these by usages of  `GTI.getSequentialElementStride(DL)`, 
which is a new helper function added in this PR.

This changes behavior for GEPs into vectors with element types for which the
(bit) size and alloc size is different. This includes two cases:

* Types with a bit size that is not a multiple of a byte, e.g. i1.
  GEPs into such vectors are questionable to begin with, as some elements
  are not even addressable.
* Overaligned types, e.g. i16 with 32-bit alignment.

Existing tests are unaffected, but a miscompilation of a new precommitted test is fixed.

### CODE REVIEW

**Comment by phoebewang** [comment]
Can this solve https://github.com/llvm/llvm-project/issues/68566 too?

**Comment by jasilvanus** [comment]
> Can this solve #68566 too?

I don't think it solves it, as it only fixes offset computations within GEPs and doesn't teach code in general about the correct vector layout. However, it is reducing the amount of code assuming the wrong layout.

There seem to be some clang test failures though that I'm currently looking into.

**Comment by jasilvanus** [comment]
> There seem to be some clang test failures though that I'm currently looking into.

Solved: The clang failures were caused by me having `dxv` in `$PATH`, which implicitly enabled some extra `dxv`-based tests which also fail without this patch for me. 

Clang was invoking `dxv - -o -`, and `dxv` complained about `-` not being a file. Not sure whether automatically enabling extra tests based on `$PATH` is desirable behavior, but that's a different topic.

File: llvm/test/Transforms/InstCombine/getelementptr.ll [hunk 1]
```diff
@@ -111,6 +111,20 @@ define void @test_evaluate_gep_as_ptrs_array(ptr addrspace(2) %B) {
   ret void
 }
 
+define void @test_overaligned_vec(i8 %B) {
+        ; This should be turned into a constexpr instead of being an instruction
+; CHECK-LABEL: @test_overaligned_vec(
+; TODO: In this test case, half is overaligned to 32 bits.
+;       Vectors are bit-packed and don't respect alignment.
+;       Thus, the byte offset of the second half in <2 x half> is 2 bytes, not 4 bytes:
```
Comment (nikic) [inline llvm/test/Transforms/InstCombine/getelementptr.ll]: Please do not interleave comments and CHECK lines. Move both comments outside the function.

File: llvm/include/llvm/IR/GetElementPtrTypeIterator.h [hunk 2]
```diff
@@ -108,7 +143,23 @@ class generic_gep_type_iterator {
...
+  // and the containing aggregate type, as vectors always tightly bit-pack their
+  // elements.
+  TypeSize getSequentialElementStride(const DataLayout &DL) const {
+    assert(isSequential());
+    Type *ElemTy = getIndexedType();
+    TypeSize ElemSizeInBits = isVector() ? DL.getTypeSizeInBits(ElemTy)
+                                         : DL.getTypeAllocSizeInBits(ElemTy);
+    // Check for invalid GEPs that are not byte-addressable.
+    assert(ElemSizeInBits.isKnownMultipleOf(8));
+    return ElemSizeInBits.divideCoefficientBy(8);
```
Comment (nikic) [inline llvm/include/llvm/IR/GetElementPtrTypeIterator.h:161]:
```suggestion
    if (isVector()) {
      assert(DL.typeSizeEqualsStoreSize(ElemTy) && "Not byte-addressable");
      return DL.getTypeStoreSize(ElemTy);
    }
    return DL.getTypeAllocSize(ElemTy);
```
Avoid explicit `8` constants.

**Review by nikic — COMMENTED** [review]
Looks fine to me.

Alternative would be to forbid GEP indexing into vectors entirely.

File: llvm/test/Transforms/InstCombine/getelementptr.ll (hunk 1 above)
Comment (jasilvanus) [reply llvm/test/Transforms/InstCombine/getelementptr.ll]: Thanks, I moved the first comment away. The TODO is already removed in the main commit of this PR.

**Comment by jasilvanus** [comment]
> Alternative would be to forbid GEP indexing into vectors entirely.

I agree that it would be better if there just were no vector GEPs, but that breaks importing older modules, and that cannot be easily auto-upgraded (convert to byte-GEPs?). Even worse, DXIL uses different vector semantics where alignment in vectors is respected, so importing DXIL already is a challenge, but if vector GEPs in DXIL were to be replaced by byte-GEPs as part of auto-upgrade, preserving DXIL semantics becomes difficult.

**Comment by jasilvanus** [comment]
I'm now on vacation for two weeks, returning Jan 4th.
I'll merge the PR then (if it is approved by then).

**Review by nikic — APPROVED** [review]
LGTM

I think this is a reasonable change in any case.

**Comment by nunoplopes** [comment]
Alive2 is complaining about one of the tests:
```llvm
@Global = global 10 bytes, align 1

define void @test_overaligned_vec(i8 %B) {
  %A = gep ptr @Global, 4 x i64 0, 4 x i64 1
  store i8 %B, ptr %A, align 1
  ret void
}
=>
@Global = global 10 bytes, align 1

define void @test_overaligned_vec(i8 %B) {
  %__constexpr_0 = gep inbounds ptr @Global, 10 x i64 0, 1 x i64 2
  store i8 %B, ptr %__constexpr_0, align 1
  ret void
}
Transformation doesn't verify! (unsound)
ERROR: Mismatch in memory

Example:
i8 %B = #x11 (17)

Source:
ptr %A = pointer(non-local, block_id=0, offset=4)
```

GEP of 4 vs 2 bytes.

**Comment by jasilvanus** [comment]
Interesting, here it is online: https://alive2.llvm.org/ce/z/Bm5gP2
The complaint goes away if `half` is naturally aligned (`f16:16`).

Looks like an Alive2 bug with respect to vector bit layout?

**Pull request merged**